src/twine/ppx/imandrakit_twine_ppx.ml

```diff
@@ -0,0 +1,807 @@
+            let def = decode_expr_of_tydecl ty in
+            A.Str.eval
+              [%expr
+                [%e A.Exp.ident (lid ~loc @@ deser_name_ref_of_ty ty)]
+                := [%e def]]
+          in
+          [ str_encode; str_decode ]
+        ))
+      type_declarations
+  in
+
+  let bracket_warn stri =
+    let loc =
+      match type_declarations with
+      | [] -> Location.none
+      | ty :: _ -> ty.ptype_loc
+    in
+    let disable = [%stri [@@@ocaml.warning "-27-33-39"]] in
+    let enable = [%stri [@@@ocaml.warning "+27+33+39"]] in
+    (disable :: stri) @ [ enable ]
+  in
+  bracket_warn
+    (List.flatten
+       [
+         forward_defs_refs;
+         defs_forward;
+         ser_defs_poly;
+         deser_defs_poly;
+         defs_init;
+       ])
+
+let generate_impl ~ctxt:_ (rec_flag, type_declarations) =
+  try generate_impl_ (rec_flag, type_declarations)
+  with Error_gen (loc, msg) ->
+    (* emit an error in generated code *)
+    let str0 =
+      [%stri let () = [%error [%e A.Exp.constant (A.Const.string msg)]]]
+    in
+    [ str0 ]
+
+let generate_intf_ type_declarations =
+  let values =
+    List.map
+      (fun ty ->
+        let loc = ty.ptype_loc in
+        let poly_sers =
+          param_names ty
+          |> List.map (fun n ->
+                 [%type: [%t A.Typ.var n] Imandrakit_twine.Encode.encoder])
+        and poly_desers =
+          param_names ty
+          |> List.map (fun n ->
+                 [%type: [%t A.Typ.var n] Imandrakit_twine.Decode.decoder])
+        in
+
+        (* type expression for [ty], like [(int, bool) mypair] *)
+        let tye =
+          A.Typ.constr ~loc (lid_of_str ty.ptype_name)
+            (List.map A.Typ.var @@ param_names ty)
+        in
+
+        (* declare top function *)
+        let name_ser = ser_name_of_ty ty in
+        let name_deser = deser_name_of_ty ty in
+        let decl_ser =
+          let ty =
+            mk_arrow ~loc poly_sers
+            @@ [%type: [%t tye] Imandrakit_twine.Encode.encoder]
+          in
+          A.Val.mk { loc; txt = name_ser } ty
+        and decl_deser =
+          let ty =
+            mk_arrow ~loc poly_desers
+            @@ [%type: [%t tye] Imandrakit_twine.Decode.decoder]
+          in
+          A.Val.mk { loc; txt = name_deser } ty
+        in
+        [ decl_ser; decl_deser ])
+      type_declarations
+    |> List.flatten
+  in
+  List.map (fun s -> A.Sig.value s) values
+
+let generate_intf ~ctxt:_ (_rec_flag, type_declarations) =
+  try generate_intf_ type_declarations
+  with Error_gen (loc, msg) ->
+    (* emit an error in generated code *)
+    let s =
+      [%sigi: val _bad_ser : [%error [%e A.Exp.constant (A.Const.string msg)]]]
+    in
+    [ s ]
+
+let impl_generator = Deriving.Generator.V2.make_noarg generate_impl
+let intf_generator = Deriving.Generator.V2.make_noarg generate_intf
+
+let myderiver =
+  Deriving.add "twine" ~sig_type_decl:intf_generator
+    ~str_type_decl:impl_generator
```

Comment:
yeah kind of, ppx is annoying.. and we do non trivial stuff too, like declaring references for monomorphic types and forwarding to them, etc. I'm sure it could be a bit shorter, that said.